File: src/com/yx/plug/markdowntableofcontents/GenerateAnchorAction.java
package com.yx.plug.markdowntableofcontents;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.SelectionModel;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;

public class GenerateAnchorAction extends AnAction
{
    @Override
    public void update(AnActionEvent e)
    {
        VirtualFile file = e.getData(CommonDataKeys.VIRTUAL_FILE);

        boolean visible = file != null && file.getName().toLowerCase().endsWith(".md");

        e.getPresentation().setEnabledAndVisible(visible);
    }

    @Override
    public void actionPerformed(AnActionEvent anActionEvent)
    {
        //Get all the required data from data keys
        final Editor editor = anActionEvent.getRequiredData(CommonDataKeys.EDITOR);
        final Project project = anActionEvent.getRequiredData(CommonDataKeys.PROJECT);
        //Access document, caret, and selection
        final Document document = editor.getDocument();
        final SelectionModel selectionModel = editor.getSelectionModel();

        final int start = selectionModel.getSelectionStart();
        final int end = selectionModel.getSelectionEnd();
        //New instance of Runnable to make a replacement
        Runnable runnable = new Runnable()
        {
            @Override
            public void run()
            {
                document.replaceString(start, end, "Replacement1231231");
            }
        };
        //Making the replacement
        WriteCommandAction.runWriteCommandAction(project, runnable);
        selectionModel.removeSelection();
    }
}

File: src/com/yx/plug/markdowntableofcontents/GenerateAnchorAction.kt
package com.yx.plug.markdowntableofcontents

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.editor.Document
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.event.DocumentEvent
import com.intellij.openapi.editor.event.DocumentListener
import com.intellij.openapi.project.Project
import java.io.BufferedReader
import java.io.IOException
import java.io.StringReader
import java.util.*


class GenerateAnchorAction : BaseSelectionAction() {

    override fun actionPerformed(anActionEvent: AnActionEvent?) {

        anActionEvent ?: return

        val editor = anActionEvent.getRequiredData(CommonDataKeys.EDITOR)

        val project = anActionEvent.getRequiredData(CommonDataKeys.PROJECT)

        val headers = findHeaders(editor)

        addAnchor(project, editor.document, headers.iterator())
    }

    fun addAnchor(project: Project, document: Document, iterator: MutableIterator<Header>) {

        if (!iterator.hasNext()) {
            return
        }

        val header = iterator.next()

        val index = findIndexOfHeader(document.text, header)

        if (index <= 0) {

            addAnchor(project, document, iterator)

            return
        }

        document.addDocumentListener(object : DocumentListener {
            override fun beforeDocumentChange(event: DocumentEvent?) {
                super.beforeDocumentChange(event)
            }

            override fun documentChanged(event: DocumentEvent?) {
                super.documentChanged(event)

                document.removeDocumentListener(this)

                ApplicationManager.getApplication().invokeLater(Runnable { addAnchor(project, document, iterator) })
            }
        })

        WriteCommandAction.runWriteCommandAction(project) { document.insertString(index, "\n" + buildAnchor(header)) }
    }

    fun buildAnchor(header: Header): String {
        return "<a name=\"${header.anchor}\"></a>"
    }

    fun findIndexOfHeader(content: String, header: Header): Int {
        var bufferedReader: BufferedReader? = null

        try {
            bufferedReader = BufferedReader(StringReader(content))

            var index = -1

            for (line in bufferedReader.lines()) {

                if (line.trim().startsWith("#") && line.contains(header.title)) {
                    return index
                }

                index += line.length + 1
            }

        } catch (e: Exception) {
            println(e)
        } finally {
            try {
                bufferedReader!!.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }

        return -1
    }
}


fun findHeaders(editor: Editor): MutableList<Header> {
    val document = editor.document
    val selectionModel = editor.selectionModel

    val start = selectionModel.selectionStart
    val end = selectionModel.selectionEnd

    val fileStr = document.text

    val selectionStr = fileStr.substring(start, end)

    var bufferedReader: BufferedReader? = null

    val headers: MutableList<Header> = mutableListOf()

    try {

        bufferedReader = BufferedReader(StringReader(selectionStr))

        for (line in bufferedReader.lines()) {
            var newLine: String = line.trim()

            if (!newLine.startsWith("- ") || !newLine.contains("#")) {
                continue
            }

            newLine = newLine.replace("- ", "").replace("[", "").replace("]", "")
                    .replace("(", "").replace(")", "")

            var arr = newLine.split("#")

            if (arr.size != 2) {
                continue
            }

            headers.add(Header(arr[0].trim(), arr[1].trim()))
        }

    } catch (e: Exception) {
        println(e)
    } finally {
        try {
            bufferedReader!!.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    return headers
}
